### PR TITLE
Cached IBlockData to CraftBlockData conversion

### PR DESCRIPTION
**Pull request by hafarooki**
Noticed how inefficient this was, so ~~Aikar~~ I made a solution for it by using functions instead *cough*

### CODE REVIEW

File: Spigot-Server-Patches/0348-Replace-inefficient-reflection-for-Block-getBlockDat.patch [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements IBlockData {
+ 
+@@ -11,4 +12,14 @@ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements I
+     public Block getBlock() {
+         return (Block) this.e_;
+     }
++
++    // Paper start - cache craftblockdata creation
++
++    private com.google.common.base.Function<IBlockData, CraftBlockData> wrapperFunction = CraftBlockData.getWrapperFunction(this);
```
Comment (aikar) [inline Spigot-Server-Patches/0348-Replace-inefficient-reflection-for-Block-getBlockDat.patch]: use real function class

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0348-Replace-inefficient-reflection-for-Block-getBlockDat.patch [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
++
++    @Override
++    public CraftBlockData getBlockData() {
++        return wrapperFunction == null ? null : wrapperFunction.apply(this);
```
Comment (aikar) [inline Spigot-Server-Patches/0348-Replace-inefficient-reflection-for-Block-getBlockDat.patch]: if null, it should return a new CraftBlockData

**Review by aikar — CHANGES_REQUESTED** [review]

**Comment by hafarooki** [comment]
how about now

**Comment by hafarooki** [comment]
fixed the null thing

**Comment by hafarooki** [comment]
**EDIT: I'm retarded disregard**

Tested it.. and it makes it slower?
Code (too lazy to JUnit): https://gist.github.com/Miclebrick/aab6c56b949f1095a3e070896cdf6e26
With latest regular paper:
https://gist.github.com/Miclebrick/bd995c2b61242b8423c26383996659e6
With my patch:
https://gist.github.com/Miclebrick/a43d4e6a96efcfec54f3bc8211acbb8e

It works either way, but it's more than twice as slow without reflection somehow. Why?

**EDIT: I'm retarded disregard**

**Comment by hafarooki** [comment]
Ok my method of benchmarking was way too lazy.
Here's a better test.
code: 
https://gist.github.com/Miclebrick/66bfea74ab00cccbd933d4bbd3fe5e17
With patch:
```
[08:26:17 INFO]: [Erudition] Warmup Time: 2,666,717,557
[08:26:17 INFO]: [Erudition] Elapsed: 10,423,223
```

Without patch:
```
[08:28:05 INFO]: [Erudition] Warmup Time: 16,620,594,612
[08:28:05 INFO]: [Erudition] Elapsed: 86,348,788
```

(commas added) (time in nanos)

Watchdog freaked the heck out with the second one

**Comment by hafarooki** [comment]
But yeah it works functionally and it's 8x faster <3

**Comment by Black-Hole** [comment]
How about caching not just the class but instances with the right properties set and call clone() on these?

**Comment by hafarooki** [comment]
It caches the method to create it. I had that idea too, but I don't know if it'd really be much more efficient.... I guess it would for complicated states. That'd fit more with my original PR which didn't work because API BlockData is apparently mutable even though NMS BlockData isn't.

**Comment by hafarooki** [comment]
Actually, apparently it is already cloneable. I'll make it clone instead later today

**Comment by Miclebrick** [comment]
Updated it, going to test it now

**Comment by hafarooki** [comment]
There we go, I redid the patch, it uses cloning now. Did some testing with the plugin I was working on when I discovered this performance issue. Before, CraftBlockData#fromData was using by far the most time in my sampler, now it says 0.00%.

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+     public Block getBlock() {
+         return (Block) this.e_;
+     }
++
++    // Paper start - impl cached craft block data, lazy load to fix issue with loading at the wrong time
++    private CraftBlockData cachedCraftBlockData;
++
++    @Override
++    public CraftBlockData getCachedCraftBlockData() {
++        if(cachedCraftBlockData == null) cachedCraftBlockData = java.util.Objects.requireNonNull(CraftBlockData.createData(this));
```
Comment (stephan-gh) [inline Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: When would `createData` return null?

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     }
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+index 472e89c43..705f8d08e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+@@ -454,6 +454,7 @@ public class CraftBlockData implements BlockData {
+         register(net.minecraft.server.BlockWitherSkull.class, org.bukkit.craftbukkit.block.impl.CraftWitherSkull.class);
+         register(net.minecraft.server.BlockWitherSkullWall.class, org.bukkit.craftbukkit.block.impl.CraftWitherSkullWall.class);
+         register(net.minecraft.server.BlockWoodButton.class, org.bukkit.craftbukkit.block.impl.CraftWoodButton.class);
++        Block.REGISTRY_ID.iterator().forEachRemaining(IBlockData::getCachedCraftBlockData); // Paper - initialize cached data for all IBlockData instances after registration
```
Comment (stephan-gh) [inline Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: Is this even necessary? Couldn't you just let it be lazily initialized the first time it is needed?

Comment (hafarooki) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: That'd be slower when plugins are calling it during the run of the server, no reason *not* to, the startup impact is insignificant. Also, if it fails for some reason, it fails at startup

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch (hunk 3 above)
Comment (hafarooki) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: Dunno, but I had some weird issues and if it happens again this way it'll show up earlier and be easier to debug

Comment (stephan-gh) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: Well, as far as I can tell it is impossible that the method returns null, so this seems redundant.

Comment (hafarooki) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: fine I'll remove it :P

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch (hunk 4 above)
Comment (aikar) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: why even do this? Initialize it in the constructor of BlockData like you did the previous pass.

Comment (hafarooki) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch]: see discord for reasoning that broke it

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ 
+ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements IBlockData {
+ 
+@@ -11,4 +12,14 @@ public class BlockData extends BlockDataAbstract<Block, IBlockData> implements I
+     public Block getBlock() {
+         return (Block) this.e_;
+     }
++
++    // Paper start - impl cached craft block data, lazy load to fix issue with loading at the wrong time
++    private CraftBlockData cachedCraftBlockData;
```
Comment (aikar) [inline Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch:25]: nevermind

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch (hunk 5 above)
Comment (hafarooki) [reply Spigot-Server-Patches/0348-Cached-IBlockData-to-CraftBlockData-conversion.patch:25]: that breaks it

**Review by aikar — APPROVED** [review]

**Comment by hafarooki** [comment]
~~yay now it's my code instead of something aikar told me to copy paste to replace my initial pr~~

**Pull request merged**